---
title: "Unpaired Variant Consensus Workflow"
author: "Amy Paguirigan"
date: "03/06/2020"
output:
  pdf_document: 
    toc: yes
editor_options: 
  chunk_output_type: console
---
# Packages and Credential Management
```{r eval=FALSE, include=FALSE}
require(remotes)
remotes::install_github('FredHutch/fh.wdlR')
remotes::install_github('FredHutch/tgR')
```

## Load Packages
```{r}
library(fh.wdlR); library(tidyverse); library(aws.s3); library(tgR); library(httr); 
```

## Set Credentials (run one line)
```{r}
setCreds(tokenSet = "file", path = "~/github/cred/paguirigan.R") 
```


# Analysis Workflow
## Choose a Workflow
Workflow is:  PairedFastqsToUnmappedBam
Download the current master branch copy and unzip the file. 
```{r}
workflowDir <- "./Workflows/newSequencing/"
repoName <- "FredHutch/tg-wdl-unpairedVariantCaller"

this <- httr::GET(url = paste0("https://github.com/", repoName ,"/archive/master.zip"),
    httr::authenticate(Sys.getenv("GITUSERNAME"), Sys.getenv("GITTOKEN")),
    httr::write_disk(paste0(workflowDir, "master.zip"), overwrite = TRUE))

unzip(zipfile = "./Workflows/newSequencing/master.zip",
      exdir = "./Workflows/newSequencing/")
file.remove("./Workflows/newSequencing/master.zip")

list.dirs(path = "./Workflows/newSequencing/")
workflowFiles <- "./Workflows/newSequencing/tg-wdl-unpairedVariantCaller/gizmo/"
```

# Make a Manifest
In this workflow, the input data must be unmapped bams with quality control checks.  Also, to run the workflow you'll need the bed file for the dataset up in S3.
```{r}
subData <- monsterMash %>% filter(workflowID == "fastq-unmappedbam") %>% dropWhen()

unpaired <- subData %>% select(molecular_id, omics_sample_name, key, seq_libtype)
unpaired$bamLocation <- paste0("s3://fh-pi-paguirigan-a/", unpaired$key)

unpaired$bedLocation <- ifelse(unpaired$seq_libtype == "one", "s3://fh-ctr-public-reference-data/reagent_specific_data/sequencing_panel_bed/TruSight-One-Illumina/TruSight_One_v1.1-liftover-hg38.bed",
                               ifelse(unpaired$seq_libtype == "tsm", "s3://fh-ctr-public-reference-data/reagent_specific_data/sequencing_panel_bed/TruSight-Myeloid-Illumina/trusight-myeloid-amplicon-v4-track_interval-liftOver-hg38.bed",
                                      ifelse(unpaired$seq_libtype == "washu", "s3://fh-ctr-public-reference-data/reagent_specific_data/sequencing_panel_bed/WASHU-AML/washU_intervals-remapped-hg38.bed",
                                             ifelse(unpaired$seq_libtype == "agssex6", "s3://fh-ctr-public-reference-data/reagent_specific_data/sequencing_panel_bed/AgilentSureSelectv6-Exome/S07604514_Covered-remapped-hg38.bed",
                                                    ifelse(unpaired$seq_libtype == "agssex5", "s3://fh-ctr-public-reference-data/reagent_specific_data/sequencing_panel_bed/AgilentSureSelectv5-Exome/S04380110_Covered-remapped-hg38.bed",
                                                           ifelse(unpaired$seq_libtype == "tapestriAMLPanel", "s3://fh-ctr-public-reference-data/reagent_specific_data/sequencing_panel_bed/MissionBio-AMLPanel/2019-02-13-MissionBioAML.bed", ""))))))
preselect <- unpaired
preselect <- unpaired %>% filter(!is.na(bedLocation)) %>% filter(seq_libtype == "one")
manifest <- preselect %>% select(omics_sample_name, bamLocation, bedLocation, molecular_id)
```

## Name and Ship Batch File
```{r}
batchFileName <- paste0("cromwell-manifests/",format(Sys.Date(), "%Y-%m-%d-"), "HALOValidation.tsv")
paste0("s3://fh-pi-paguirigan-a-genomicsrepo/", batchFileName)
s3write_using(manifest,
              FUN = write.table, quote = F, row.names = F, sep = "\t",
              object = batchFileName,
              bucket = "fh-pi-paguirigan-a-genomicsrepo")
thisManifest <- "cromwell-manifests/2020-03-06-HALOValidation.tsv"
```


# Cromwell Environment

## Create Cromwell Server
Create a Cromwell server on Gizmo or interact with a specific Cromwell server.  
```{r}
cromwellCreate(FredHutchId = "apaguiri", port = "1981",
        pathToServerLogs =
          "~/cromwell/tgr/server-logs/frank-%A.txt",
        pathToServerScript = 
          "~/cromwell/tgr/beAllEndAll.sh",
        pathToParams = 
          "~/cromwell/tgr/allTheParams.sh")
```

## Set Cromwell Environment 
```{r}
setCromwellURL(FredHutchId = "apaguiri", jobId = "...", port = "2020")
Sys.getenv("CROMWELLURL")
Sys.setenv("CROMWELLURL" = "http://...")
```

## Validate and Submit Job
```{r}
list.files(path = workflowFiles)
womtoolValidate(paste0(workflowFiles, "unpairedVariants.wdl"))

thisJob <- cromwellSubmitBatch(
  WDL = paste0(workflowFiles, "unpairedVariants.wdl"),
  Batch = paste0(workflowFiles, "unpairedVariants-batch.json"),
  Params = paste0(workflowFiles, "unpairedVariants-parameters.json"),
  Options = paste0(workflowDir, "local-options.json"),
  Labels = data.frame("workflowType" = "haloValidation-correctedMemory"))

thisOne <- thisJob$id; thisOne
```

# Monitor Running Jobs

## Attempts
Jobs of interest
```{r}
thisOne <- ".." 
jobs <- cromwellJobs()
```

## Workflow level metadata
```{r}
w <- cromwellWorkflow(thisOne); w$status
WTF <- cromwellGlob(thisOne); WTF[["failures"]]
```

## Call level metadata 
```{r}
c <- cromwellCall(thisOne); c %>% group_by(executionStatus, callName) %>% summarize(status = n())
ca <- cromwellCache(thisOne); ca %>% group_by(callCaching.hit, callName) %>% summarize(hits = n())

butWhy <- left_join(cromwellCall(thisOne) %>% mutate_all(as.character), 
                    cromwellCache(thisOne) %>% mutate_all(as.character)); butWhy %>% group_by(callName, executionStatus, callCaching.hit) %>% summarize(hits = n()) %>% arrange(desc(executionStatus))
```

## Failure metadata and abort
```{r}
f <- cromwellFailures(thisOne); f
#abort <- cromwellAbort(thisOne) # Careful with this
```

# Output Processing
## Find workflow outputs
```{r}
out <- cromwellOutputs(thisOne) 
batchFile <- s3read_using(FUN = read.delim, stringsAsFactors = F,
                           object = thisManifest,
                           bucket = "fh-pi-paguirigan-a-genomicsrepo")
batchFile$shardIndex <- as.character(seq(from = 0, to = nrow(batchFile)-1, by = 1))
annotatedOutputs <- inner_join(batchFile, out, by = c("shardIndex")) # if no molecular_id
```

